ql/src/java/org/apache/hadoop/hive/ql/queryhistory/QueryHistoryService.java

```diff
@@ -0,0 +1,248 @@
+    } else {
+      LOG.info("Query History max queue memory byte size is {}: records will be persisted when the queue consumes " +
+              "more than {} bytes in the memory",
+          maxQueueSizeInMemoryBytes, maxQueueSizeInMemoryBytes);
+    }
+  }
+
+  protected QueryHistoryRepository createRepository(HiveConf conf) {
+    initializeSessionForTableCreation(conf);
+
+    try {
+      return (QueryHistoryRepository) ReflectionUtil.newInstance(
+          conf.getClassByName(conf.get(HiveConf.ConfVars.HIVE_QUERY_HISTORY_REPOSITORY_CLASS.varname)), conf);
+    } catch (ClassNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private void initializeSessionForTableCreation(HiveConf conf) {
+    HiveConf.setVar(conf, HiveConf.ConfVars.HIVE_QUERY_ID, QueryHistoryRepository.QUERY_ID_FOR_TABLE_CREATION);
+    SessionState ss = SessionState.get(); // if there is a SessionState in the called thread, we can use that
+    if (ss == null){
+      ss = SessionState.start(conf);
+    }
+    ss.addQueryState(QueryHistoryRepository.QUERY_ID_FOR_TABLE_CREATION,
+        new QueryState.Builder().withHiveConf(conf).build());
+  }
+
+  /**
+   * Creates a Query History record from a given DriverContext and adds it to the queue.
+   * This is done asynchronously to completely decouple from the threads that run the query to make this service work
+   * silently in the background.
+   * @param driverContext a DriverContext containing all the needed information for a record to be created
+   */
+  public void handleQuery(DriverContext driverContext) {
+    long recordCreateStart = Time.monotonicNow();
+    QueryHistoryRecord record = createRecord(driverContext);
+    long recordCreateEnd = Time.monotonicNow();
+    queryHistoryQueue.add(record);
+    LOG.info("Created query history record in {}ms, current queue size: {}",
+        recordCreateEnd - recordCreateStart, queryHistoryQueue.size());
+
+    persistRecordsIfNeeded();
+  }
+
+  private void persistRecordsIfNeeded() {
+    if (needToPersist()) {
+      persistWithBatchSize(maxBatchSize);
+    } else {
+      LOG.debug("Not persisting history records yet, current queue size: {}, maxBatchSize: {}",
+          queryHistoryQueue.size(), maxBatchSize);
+    }
+  }
+
+  private void persistWithBatchSize(int batchSize) {
+    CountDownLatch latch = new CountDownLatch(1);
+    LOG.debug("Submitting a job to persist {} history records", queryHistoryQueue.size());
+    persistExecutor.submit(() -> {
+      repository.persist(queryHistoryQueue);
+      latch.countDown();
+    });
+    if (batchSize == 0) {
+      waitForAllRecordsToBePersisted(latch);
+    }
+  }
+
+  private void waitForAllRecordsToBePersisted(CountDownLatch queueEmptyCondition) {
+    try {
+      queueEmptyCondition.await();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(e);
+    }
+  }
+
+  @VisibleForTesting
+  boolean needToPersist() {
```

Comment:
should it be synchronized or something? What if multiple query sessions call `handleQuery` and we need to persist?

Reply:
makes sense, I think it doesn't matter what decision is made in this method and whether it's made synchronized, the actual problem can happen while flushing the queue, so I think instead of:
```
CompletableFuture.runAsync(
        () -> repository.flush(queryHistoryQueue), flushExecutor);
```

we might want to use:
```
CompletableFuture.runAsync(
        () -> {
          synchronized (QueryHistoryService.this) {
            repository.flush(queryHistoryQueue);
          }
        }, flushExecutor);
```

Reply:
oh, wait, we have just 1 flush thread, so no need for synchronize

Reply:
right! and that's by design, so I'm removing synchronized